File: Source/Old C Code/grader.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>

#include "../custom/custom.h"
#include "grader.h"

int grader()
{
    clear();
    
	//variables used for directory reading
	DIR* FD;
    struct dirent* in_file;
    FILE* master_results;
    FILE* current_file;
    char buffer[250];
    FILE* quiz;

    //track number of student files ran through
    int numStudents=0;

    //get name of quiz to run
    char quizName[140]="quizes/";
    char temp[100];
    printf("What quiz do you want to grade(100 char max): ");
    //scanf("%s", temp);
    fflush(stdin);
    fgets(temp, 100, stdin);
    line();
    strcat(quizName, temp);

    //check if choice is empty
    if(strcmp(temp, "\n") == 0)
    {
        printf("Quiz name cannot be empty!\n");
        pete();
        return 1;
    }

    //add .quizfile extension to string
    strtok(quizName, "\n");
    strcat(quizName, ".quizfile");

    //opening quiz file stream r
    quiz = fopen(quizName, "r+");
    if(quiz==NULL)
    {
        printf("There was an error opening the quiz file.\n");
        printf("Press ENTER to close Grader.\n");
        fflush(stdin);
        getchar();
        return 1;
    }

    //get number of questions in quiz
    int numQs;
    char buffer2[225];
    fseek(quiz, 3, SEEK_SET);
    fgets(buffer2, 225, quiz);
    numQs = atoi(buffer2);

    //create array to store question statistics and fill with 0s
    float questionStats[4][numQs+1];
    int i, j;
    for(i=0;i<4;i++)
    {
        for(j=0;j<numQs+1;j++)
        {
            questionStats[i][j]=0;
        }
    }

    //create grades directory
    make_directory("grades");

	//create a grades/quizname/overall.grade file to store calculations
    char gradeFile[300]="grades/";
    strcat(gradeFile, temp);
    strtok(gradeFile, "\n");
    strcat(gradeFile, "/");
    make_directory(gradeFile);
    strcat(gradeFile, "overall.grade");
    master_results = fopen(gradeFile, "w");
    if (master_results == NULL)
    {
        fprintf(stderr, "Error: Failed to create %s\n", gradeFile);
        printf("Press ENTER to close Grader.\n");
        fflush(stdin);
        getchar();
        return 1;
    }

    //create a quizDump/quizName.dump file to store the test
    FILE* dumpFP;
    make_directory("quizDump");
    char dumpFile[300]="quizDump/";
    strcat(dumpFile, temp);
    strtok(dumpFile, "\n");
    strcat(dumpFile, ".dump");
    dumpFP = fopen(dumpFile, "w");
    if (dumpFP == NULL)
    {
        fprintf(stderr, "Error: Failed to create %s\n", dumpFile);
        printf("Press ENTER to close Grader.\n");
        fflush(stdin);
        getchar();
        return 1;
    }

    //creating string for results directory
    char resultsDir[300]="results/";
    strcat(resultsDir, temp);
    strtok(resultsDir, "\n");
    strcat(resultsDir, "/");

    //checking if results-quizname/ directory can open and opening if it exists
    if (NULL == (FD = opendir (resultsDir))) 
    {
        printf("Error: Failed to open %s directory\n", resultsDir);
        printf("This error is common if no one has taken the selected quiz yet.\n");
        printf("Press ENTER to close Grader.\n");
        fflush(stdin);
        getchar();
        fclose(master_results);
        return 1;
    }

    //looping through all files in results/ directory
    while ((in_file = readdir(FD))) 
    {
    	//remove parent directory listings
        if (!strcmp (in_file->d_name, "."))
        {
            continue; 
        }
        if (!strcmp (in_file->d_name, ".."))    
        {
            continue; //do not print
        }

        //get name of file with results/ dir prepended
        char fileString[150];
        strcpy(fileString, resultsDir);
        strcat(fileString, in_file->d_name);

        //open file in r+ mode and check
        current_file = fopen(fileString, "r+");
        if (current_file == NULL)
        {
            printf("Error: Failed to open file %s!\n", in_file->d_name);
            printf("Press ENTER to close Grader.\n");
            fflush(stdin);
            getchar();
            fclose(master_results);
            return 1;
        }
        else
        {
            printf("Now grading %s!\n", in_file->d_name);
        }

        //storing studentName for future use
        char studentName[150];
        strcpy(studentName, in_file->d_name);
        int studentNameLen = strlen(studentName);
        studentName[studentNameLen-12]='\0';

        //opening filestream for individual results
        char studentID[300]="grades/";
        strcat(studentID, temp);
        strtok(studentID, "\n");
        strcat(studentID, "/");
        strcat(studentID, in_file->d_name);
        int studentIDLen = strlen(studentID);
        studentID[studentIDLen-11] = '\0';
        strcat(studentID, "grade");
        FILE* lone_results;
        lone_results = fopen(studentID, "w+");
        if(lone_results == NULL)
        {
            printf("Error: Failed to create %s\n", studentID);
            printf("Press ENTER to close Grader.\n");
            fflush(stdin);
            getchar();
            fclose(lone_results);
            fclose(master_results);
            return 1;
        }

        //trim B: line off results file
        int cPos = ftell(current_file);
        fgets(buffer, 250, current_file);

        //declaring persistent counter variables
        int studentCorrect=0;
        float studentAverage=0;
        int questionNum=1;

        //loop through entire file and print to screen
        while (fgets(buffer, 250, current_file) != NULL)
        {
            //get question from quiz file
            char quizbuffer[300];
            while(strstr(quizbuffer, "Q: ") == NULL && strstr(quizbuffer, "EEE") == NULL)
            {
                fgets(quizbuffer, 250, quiz);  
            }
            if(strstr(quizbuffer, "Q: ") != NULL)
            {
                char *truebuffer = quizbuffer +3;
                fprintf(dumpFP, "--------------------------------\n");
                fprintf(dumpFP, "Question %d:\n", questionNum);
                questionNum++;
                fprintf(dumpFP, "%s", truebuffer);
                fgets(quizbuffer, 250, quiz);
            }

            //get answers from quiz file
            while(strstr(quizbuffer, "C: ") == NULL && strstr(quizbuffer, "EEE") == NULL)
            {
                while(strstr(quizbuffer, "A. ") == NULL && strstr(quizbuffer, "EEE") == NULL)
                {
                    fgets(quizbuffer, 250, quiz);  
                }
                if(strstr(quizbuffer, "A. ") != NULL)
                {
                    char *truebuffer = quizbuffer +3;
                    fprintf(dumpFP, "\t%s", truebuffer);
                    fgets(quizbuffer, 250, quiz);
                }
            }
            
            //get question number
            if (strstr(buffer, "Q: ") != NULL)
            {
                cPos = ftell(current_file);
                fseek(current_file, cPos-4, SEEK_SET);
                int questionNum = atoi(fgets(buffer, 225, current_file));
                cPos = ftell(current_file);

                //put question number into table
                questionStats[0][questionNum] = questionNum;

                //see if answer was right or wrong
                fgets(buffer, 250, current_file);
                fseek(current_file, -4, SEEK_CUR);
                int questionStatus = atoi(fgets(buffer, 225, current_file));

                if(questionStatus==1)
                {
                    questionStats[1][questionNum]++;
                    studentCorrect++;
                    fprintf(lone_results, "\tQuestion %d: Correct!\n", questionNum);
                }
                else
                {
                    //get studentAns and expectedAns
                    fgets(buffer, 250, current_file);
                    fseek(current_file, -4, SEEK_CUR);
                    int studentAns = atoi(fgets(buffer, 250, current_file));
                    fgets(buffer, 250, current_file);
                    fseek(current_file, -4, SEEK_CUR);
                    int expectedAns = atoi(fgets(buffer, 250, current_file));
                    questionStats[2][questionNum]++;
                    fprintf(lone_results, "\tQuestion %d: Wrong! Expected %d but received %d\n", questionNum, expectedAns, studentAns);
                }
                fprintf(lone_results, "--------------------------------\n");
            } 
        }

        //print score and percent to individual file
        studentAverage = (float) studentCorrect / numQs * 100;
        fprintf(lone_results, "Score: %d out of %d\n", studentCorrect, numQs);
        fprintf(lone_results, "Percentage: %.2f%%\n", studentAverage);

        //print user and percent to overall file
        fprintf(master_results, "ID:%s - %d/%d - %.2f%%\n", studentName, studentCorrect, numQs, studentAverage);

	    //closing current file to move onto next one
        printf("Completed grading %s!\n", in_file->d_name);
        printf("Individual grade file saved to %s!\n", studentID);
        line();
	    fclose(current_file);
        fclose(lone_results);
        numStudents++;
    }

    //calculate question averages
    for(i=1;i<numQs+1;i++)
    {
        questionStats[3][i] = questionStats[1][i] / numStudents * 100;
    }

    //print questionStats to overall.grade in table format if questions <=10
    if(numQs<=10)
    {
        fprintf(master_results, "--------------------\n");
        for(i=0;i<4;i++)
        {
            switch(i)
            {
                case 0:
                    fprintf(master_results, "Question:\t"); 
                    break;
                case 1: 
                    fprintf(master_results, "Correct :\t"); 
                    break;
                case 2: 
                    fprintf(master_results, "Wrong   :\t"); 
                    break;
                case 3: 
                    fprintf(master_results, "Percent :\t"); 
                    break;
            }
            for(j=1;j<numQs+1;j++)
            {
                if(i==3)
                {
                    fprintf(master_results, "%.2f%%\t", questionStats[i][j]); 
                }
                else
                {
                    fprintf(master_results, "%.2f\t", questionStats[i][j]); 
                }
            }
            fprintf(master_results, "\n"); 
        }
    }
    else
    {
        //print questionStats to overall.grade in list format
        fprintf(master_results, "--------------------\n");
        for(i=1;i<numQs+1;i++)
        {
            fprintf(master_results, "Question: %.2f\n", questionStats[0][i]); 
            fprintf(master_results, "\tCorrect: %.2f\n", questionStats[1][i]); 
            fprintf(master_results, "\tWrong: %.2f\n", questionStats[2][i]); 
            fprintf(master_results, "\tPercent: %.2f%%\n", questionStats[3][i]); 
                    
            fprintf(master_results, "\n"); 
        }
    }
    

    //print questionStats to screen if questions <=10
    if(numQs<=10)
    {
       for(i=0;i<4;i++)
        {
            switch(i)
            {
                case 0: 
                    printf("Question:\t");
                    break;
                case 1: 
                    printf("Correct :\t"); 
                    break;
                case 2: 
                    printf("Wrong   :\t"); 
                    break;
                case 3: 
                    printf("Percent :\t"); 
                    break;
            }
            for(j=1;j<numQs+1;j++)
            {
                if(i==3)
                {
                    printf("%.2f%%\t", questionStats[i][j]);
                }
                else
                {
                    printf("%.2f\t", questionStats[i][j]);
                }
            }
            printf("\n");
        } 
    }
    else
    {
        printf("There is more than 10 questions in this quiz.\n");
        printf("Individual question results table printed to file only.\n");
    }

    //get total average
    float classAverage=0;
    for(i=1;i<=numQs;i++)
    {
    	classAverage += questionStats[3][i];
    }
    classAverage /= numQs;
    line();
    printf("Quiz average was %.2f%%.\n", classAverage);
    fprintf(master_results, "--------------------\n");
    fprintf(master_results, "Quiz average was: %.2f%%.\n", classAverage);

    line();
    printf("Overall results saved to %s!\n", gradeFile);
    printf("Test Dump saved to %s and will be visible to students!\n", dumpFile);
    printf("Press ENTER to exit Grader.\n");
    fflush(stdin);
    getchar();

    //closing streams
    fclose(dumpFP);
    fclose(master_results);
    fclose(quiz);
	return 0;
}

File: Source/modules/grader.py
# standard imports
from pathlib import Path
import os
import shutil

# custom imports
from modules.custom import *
from modules.auth import *
from modules.classes import *
from modules.takeQuiz import *

def gradeMenu(carryClass):
    carryClass = carryClass[1:]
    clear()
    # get all quizes for current class
    quizPath = os.path.join(os.getcwd(), "classes", carryClass, "quizzes")
    filesList = os.listdir(quizPath)
    quizFolders = []
    for file in filesList:
        if Path(file).is_dir:
            quizFolders.append(file)

    # list status of all quizes
    while True:
        clear()
        i = 1
        for quiz in quizFolders:
            gradedPath = Path(os.path.join(os.getcwd(), "quiz", "overall.grade"))
            try:
                open(gradedPath, "r")
                print(str(i) + ": " + quiz + " | GRADED")
            except:
                print(str(i) + ": " + quiz + " | NOT GRADED")
            i += 1

        line()
        selection = input("Selection: ")
        if(selection == "!!"):
            return
        try:
            selection = int(selection)
            if(selection >= 1 and selection <= i):
                break
            else:
                print("Invalid selection, press ENTER to retry.")
                input()
        except:
            print("Invalid selection, press ENTER to retry.")
            input()

    # get next function call from selection
    selectedQuiz = quizFolders[selection-1]
    gradedPath = Path(os.path.join(quizpath, selectedQuiz, "overall.grade"))
    selectedQuizPath = os.path.join(quizPath, selectedQuiz, str(selectedQuiz + ".quizfile"))
    selectedQuizFolder = os.path.join(quizPath, selectedQuiz)
    try:
        open(gradedPath, "r")
        print("{0} is already graded for this class.".format(selectedQuiz))
        print("Press ENTER to return to teacher menu.")
        input()
        return
    except:
        gradeQuiz(selectedQuizPath, selectedQuizFolder)
        return

def gradeQuiz(selectedQuizPath, selectedQuizFolder):
    # get generic file streams opened
    quizStream = open(selectedQuizPath, "r")
    overallStreamPath = os.path.join(selectedQuizFolder, "overall.grade")
    overallStream = open(overallStreamPath, "w")
    dumpStreamPath = os.path.join(selectedQuizFolder, "dump.dump")
    dumpStream = open(dumpStreamPath, "w")

    # get number of questions in quiz
    quizStreamBuffer = quizStream.readline()

    #get number of questions
    quizStreamBuffer = quizStream.readline()
    quizStreamBuffer = quizStreamBuffer.split(".")
    quizStreamBuffer.pop(0)
    numQs = int(quizStreamBuffer[0])

    # create array to store results while grading
    w, h = 4, numQs+1;
    questionStats = [[0 for x in range(w)] for y in range(h)] 

    # get results directory
    resultsPath = Path(os.path.join(selectedQuizFolder, "results"))
    numStudents = 0
    for file in os.listdir(resultsPath):
        filename = os.fsdecode(file)
        if(filename.endswith(".result")):
            # this loops through all .result files
            print("Now grading {0}!".format(filename))
            
            # get name of student
            studentName = os.path.basename(file)
            studentName = studentName[:-7]

            # open student grade filestream
            studentStreampath = os.path.join(selectedQuizFolder, "grades", str(studentName + ".grade"))
            studentStream = open(studentStreamPath, "w")

            # open student results filestream
            studentsResultsStreamPath = os.path.join(selectedQuizFolder, "results", str(studentName + ".result"))
            studentResultStream = open(studentResultStreamPath, "r")

            # trim B: line off results
            buffer = studentResultStream.readline()

            # declaring needed counter variables
            studentCorrect = 0
            studentAverage = 0
            questionNum = 1

            # loop until EOF
            while True:
                buffer = studentResultStream.readline()
                if not buffer: # if nothing read
                    break

                if(buffer[0] == "Q"):
                    # get question number
                    questionNum = int(buffer[3:])
                    # write to table
                    questionStats[0][questionNum] = questionNum

                elif(buffer[0] == "C"):
                    # get answer status
                    questionStatus = int(buffer[3:])
                    if(questionStatus == 1):
                        questionStats[1][questionNum] += 1
                        studentStream.write("Question {0}: Correct!\n".format(questionNum))
                        # chop off expected and actual answer lines
                        buffer = studentResultStream.readline()
                        buffer = studentResultStream.readline()
                        studentCorrect += 1
                    else:
                        # get student answer and expected answer
                        buffer = studentResultStream.readline()
                        studentAns = int(buffer[3:])
                        buffer = studentResultStream.readline()
                        expectedAns = int(buffer[3:])
                        questionStats[2][questionNum] += 1
                        studentStream.write("Question {0}: Wrong! Expected {1} but got {2}.\n".format(questionNum, expectedAns, studentAns))  
                    studentStream.write("--------------------------------\n")
                
                else:
                    pass

            #print score and percent to individual file
            studentAverage = round(studentCorrect / numQs * 100, 2)
            studentStream.write("Score: {0} out of {1}\n".format(studentCorrect, numQs))
            studentStream.write("Percentage: {0}\n".format(studentAverage))

            #print user and percent to overall file
            overallStream.write("ID:{0} - {1}/{2} - {3}%\n".format(studentName, studentCorrect, numQs, studentAverage))

            numStudents += 1

    #calculate question averages
    for i in range(1, numQs+1):
        questionStats[3][i] = round(questionStats[1][i] / numStudents * 100, 2)

    #print questionStats to overall.grade in list format
    overallStream.write("--------------------\n")
    for i in range(1, numQs+1):
        overallStream.write("Question: {0}\n".format(questionStats[0][i]))
        overallStream.write("\tCorrect: {0}\n".format(questionStats[1][i]))
        overallStream.write("\tWrong: {0}\n".format(questionStats[2][i]))
        overallStream.write("\tPercent: {0}\n\n".format(questionStats[3][i]))

    #get total average
    classAverage = 0
    for i in range (1, numQs+1):
        classAverage += questionStats[3][i]
    classAverage /= numQs
    classAverage = round(classAverage, 2)
    from modules.custom import line
    line()
    print("Quiz average was {0}%.\n".format(classAverage))
    overallStream.write("--------------------\n")
    overallStream.write("Quiz average was: {0}%.\n".format(classAverage))

    #create test dump
    answerCount = 1
    questionCount = 1
    for line in quizStream:
        if(line[0] == "Q"):
            dumpStream.write("--------------------\n")
            answerCount = 1
            dumpStream.write("Question {0}: {1}".format(questionCount, line[3:]))
            questionCount += 1
        elif(line[0] == "A"):
            dumpStream.write("\t{0}: {1}".format(answerCount, line[3:]))
            answerCount += 1


    print("Finshed grading! Press ENTER to return to teacher menu.")
    input()
    return

def gradedQuizMenu(carryClass):
    clear()
    # get all  quizes for current class
    quizPath = os.path.join(os.getcwd(), "classes", carryClass[1:], "quizzes")
    filesList = os.listdir(quizPath)
    quizFolders = []
    for file in filesList:
        if Path(file).is_dir:
            quizFolders.append(file)

    # print only graded quizes
    while True:
        clear()
        i = 1
        for quiz in quizFolders:
            gradePath = Path(os.path.join(quizPath, "quizzes", "overall.grade"))

            try:
                open(gradePath, "r")
                print(str(i) + ": " + quiz)
                i += 1
            except:
                pass

        line()
        selection = input("Selection: ")
        if(selection == "!!"):
            return
        try:
            selection = int(selection)
            if(selection >= 1 and selection <= i):
                break
            else:
                print("Invalid selection, press ENTER to retry.")
                input()
        except:
            print("Invalid selection, press ENTER to retry.")
            input()

    # menu for selected quiz
    selectedQuiz = quizFolders[selection-1]
    while True:
        clear()
        print("1. See overall result for {0}".format(selectedQuiz))
        print("2. See quiz dump for {0}".format(selectedQuiz))
        print("3. Reset quiz {0} (remove all ungraded results, all grades, and test dump".format(selectedQuiz))
        selection = input("Selection: ")
        if(selection == "!!"):
            return
        try:
            selection = int(selection)
        except:
            print("Invalid selection, press ENTER to retry.")
            input()
        if(selection == 1):
            # view overall grade
            viewer(gradePath)
            return
        elif(selection == 2):
            # view dump
            dumpPath = os.path.join(quizPath, selectedQuiz, "dump.dump")
            viewer(dumpPath)
            return
        elif(selection == 3):
            # reset quiz
            quizDir = Path(os.path.join(quizPath, selectedQuiz))
            os.chdir(quizDir)
            os.chdir("grades")
            fileList = [file for file in os.listdir(os.getcwd())]
            for file in fileList:
                os.remove(os.path.join(os.getcwd(), file))

            os.chdir("../results")
            fileList = [file for file in os.listdir(os.getcwd())]
            for file in fileList:
                os.remove(os.path.join(os.getcwd(), file))

            os.chdir("../../../../")

            dumpPath = os.path.join(quizPath, selectedQuiz, "dump.dump")
            try:
                os.remove(dumpPath)
            except:
                pass

            lockPath = os.path.join(quizPath, selectedQuiz, "lock.lock")
            try:
                os.remove(lockPath)
            except:
                pass

            overallpath = os.path.join(quizPath, selectedQuiz, "overall.grade")
            os.remove(overallPath)

            clear()
            print("Quiz reset successfully. Press ENTER to return to teacher menu.")
            input()
            return

        else:
            print("Invalid selection, press ENTER to retry.")
            input()

def deleteQuiz(carryClass):
    clear()
    # get all quizes for current class
    quizPath = os.path.join(os.getcwd(), "classes", carryClass[1:], "quizzes")
    filesList = os.listdir(quizPath)
    quizFolders = []
    for file in filesList:
        if Path(file).is_dir:
            quizFolders.append(file)

    # print only graded quizes
    while True:
        clear()
        i = 1
        for quiz in quizFolders:
            print(str(i) + ": " + quiz)
            i += 1

        line()
        selection = input("Selection: ")
        if(selection == "!!"):
            return
        try:
            selection = int(selection)
            if(selection >= 1 and selection <= i):
                break
            else:
                print("Invalid selection, press ENTER to retry.")
                input()
        except:
            print("Invalid selection, press ENTER to retry.")
            input()
    
    # confirm
    clear()
    selectedQuiz = quizFolders[selection-1]
    while True:
        print("Are you sure you want to delete {0}?".format(selectedQuiz))
        print("1. Yes")
        print("2. No")
        line()
        selection = input("Selection: ")
        if(selection == "!!"):
            return
        try:
            selection = int(selection)
            if(selection == 1 or selection == 2):
                break
            else:
                print("Invalid selection, press ENTER to retry.")
                input()
        except:
            print("Invalid selection, press ENTER to retry.")
            input()

    if(selection == 1):
        # delete quiz directory
        selectedQuizPath = os.path.join(quizPath, selectedQuiz)
        shutil.rmtree(selectedQuizPath)
        clear()
        print("Quiz {0} deleted successfully.".format(selectedQuiz))
        print("Press ENTER to return to teacher menu.")
        input()
        return
    else:
        return

def lockQuiz(carryClass):
    clear()
    # get all quizes for current class
    quizPath = os.path.join(os.getcwd(), "classes", carryClass[1:], "quizzes")
    filesList = os.listdir(quizPath)
    quizFolders = []
    for file in filesList:
        if Path(file).is_dir:
            quizFolders.append(file)

    # print lock status for all quizes
    while True:
        clear()
        i = 1
        for quiz in quizFolders:
            lockPath = Path(os.path.join(quizPath, quiz, "lock.lock"))
            try:
                open(lockPath, "r")
                print(str(i) + ": " + quiz + "| LOCKED")
            except:
                print(str(i) + ": " + quiz + "| UNLOCKED")
            i += 1

        line()
        selection = input("Selection: ")
        if(selection == "!!"):
            return
        try:
            selection = int(selection)
            if(selection >= 1 and selection <= i):
                break
            else:
                print("Invalid selection, press ENTER to retry.")
                input()
        except:
            print("Invalid selection, press ENTER to retry.")
            input()

    selectedQuiz = quizFolders[selection-1]
    # toggle lock state
    unlockPath = Path(os.path.join(quizPath, quiz, "unlock.unlock"))
    try:
        # unlock
        open(lockPath, "r")
        os.rename(lockPath, unlockPath)
        clear()
        print("Quiz {0} has been unlocked.".format(selectedQuiz))
        print("Press ENTER to return to teacher menu.")
        input()
        return
    except:
        # lock
        os.rename(unlockPath, lockPath)
        clear()
        print("Quiz {0} has been locked.".format(selectedQuiz))
        print("Press ENTER to return to teacher menu.")
        input()
        return

